packages/core/src/routes/session/passwordless.ts

```diff
@@ -148,195 +154,103 @@ export default function passwordlessRoutes<T extends AnonymousRouter>(
-      );
+    await updateUserById(id, { lastSignInAt: Date.now() });
+    await assignInteractionResults(ctx, provider, { login: { accountId: id } });
 
-      await verifyPasscode(jti, PasscodeType.SignIn, code, { phone });
-      const { id } = await findUserByPhone(phone);
-      ctx.log(type, { userId: id });
+    return next();
+  });
 
-      await updateUserById(id, { lastSignInAt: Date.now() });
-      await assignInteractionResults(ctx, provider, { login: { accountId: id } }, true);
+  router.post(`${signInRoute}/email`, async (ctx, next) => {
+    const { result } = await provider.interactionDetails(ctx.req, ctx.res);
+    const verificationStorage = parseVerificationStorage(result);
 
-      return next();
-    }
-  );
+    const { email, flow, expiresAt } = verificationStorage;
+    assertThat(email, new RequestError({ code: 'session.passwordless_not_verified', status: 401 }));
 
-  router.post(
-    `${signInRoute}/email/send-passcode`,
-    koaGuard({ body: object({ email: string().regex(emailRegEx) }) }),
-    async (ctx, next) => {
-      const { jti } = await provider.interactionDetails(ctx.req, ctx.res);
-      const { email } = ctx.guard.body;
-      const type = 'SignInEmailSendPasscode';
-      ctx.log(type, { email });
+    const type = getPasswordlessRelatedLogType(PasscodeType.SignIn, 'email');
+    ctx.log(type, { email, flow, expiresAt });
 
-      assertThat(
-        await hasUserWithEmail(email),
-        new RequestError({ code: 'user.email_not_exists', status: 422 })
-      );
+    checkVerificationSessionByFlow(PasscodeType.SignIn, verificationStorage);
 
-      const passcode = await createPasscode(jti, PasscodeType.SignIn, { email });
-      const { dbEntry } = await sendPasscode(passcode);
-      ctx.log(type, { connectorId: dbEntry.id });
-      ctx.status = 204;
+    assertThat(
+      await hasUserWithEmail(email),
+      new RequestError({ code: 'user.email_not_exists', status: 422 })
+    );
+    const { id } = await findUserByEmail(email);
+    ctx.log(type, { userId: id });
 
-      return next();
-    }
-  );
+    await updateUserById(id, { lastSignInAt: Date.now() });
+    await assignInteractionResults(ctx, provider, { login: { accountId: id } });
 
-  router.post(
-    `${signInRoute}/email/verify-passcode`,
-    koaGuard({ body: object({ email: string().regex(emailRegEx), code: string() }) }),
-    async (ctx, next) => {
-      const { jti } = await provider.interactionDetails(ctx.req, ctx.res);
-      const { email, code } = ctx.guard.body;
-      const type = 'SignInEmail';
-      ctx.log(type, { email, code });
+    return next();
+  });
 
-      assertThat(
-        await hasUserWithEmail(email),
-        new RequestError({ code: 'user.email_not_exists', status: 422 })
-      );
+  router.post(`${registerRoute}/sms`, async (ctx, next) => {
+    const { result } = await provider.interactionDetails(ctx.req, ctx.res);
+    const verificationStorage = parseVerificationStorage(result);
```

Comment:
just cross my mind, would it be better if we define multiple guards? Like what we did in schemas

smsSignInSessionGuard
emailSignInSessionGuard
smsRegisterSessionGuard
emailRegisterSessionGuard

You may guard the flow and sign-in method at the same time,  and you can infer the proper type as well.

Reply:
could even make it as a koa middleware later

Reply:
Updated.